third_party/move/tools/move-asm/src/syntax.rs

```diff
@@ -0,0 +1,777 @@
+//!     simple_type | "(" COMMA_LIST(simple_type) ")"
+//!
+//! simple_type :=
+//!   QID [ type_args ] | "(" type ")"
+//!
+//! type_args :=
+//!   "<" COMMA_LIST(type) ">"
+//!
+//! instruction :=
+//!   (INDENT | LOOKAHEAD(ID ":"))
+//!   ( "local" local | opcode_id [ COMMA_LIST(argument) ] )
+//!
+//! opcode_id := IDENT # current instr set
+//! argument :=
+//!   VALUE | QID [ type_args ] | type_args
+
+use codespan::{RawIndex, Span};
+use codespan_reporting::diagnostic::{Diagnostic, Label, Severity};
+use move_binary_format::file_format::Visibility;
+use move_core_types::{
+    ability::{Ability, AbilitySet},
+    account_address::AccountAddress,
+    identifier::Identifier,
+    language_storage::ModuleId,
+    u256::U256,
+};
+use std::{
+    collections::VecDeque,
+    fmt::{Display, Formatter},
+    ops::Range,
+    str::FromStr,
+    string::ToString,
+};
+
+// ==========================================================================================
+// Diagnostics
+
+/// Currently we use as locations span's (no filename), but keep this abstracted in case
+/// we want to change this.
+pub(crate) type Loc = Span;
+
+/// Currently we produce diagnostics for singe files without explicit name.
+pub(crate) type Diag = Diagnostic<()>;
+
+/// A result with a set of diagnostics
+pub(crate) type AsmResult<A> = Result<A, Vec<Diag>>;
+
+pub(crate) fn error(loc: Loc, message: impl ToString) -> Vec<Diag> {
+    vec![Diag::new(Severity::Error)
+        .with_labels(vec![Label::primary((), loc)])
+        .with_message(message.to_string())]
+}
+
+pub(crate) fn loc(range: Range<usize>) -> Loc {
+    Loc::new(range.start as RawIndex, range.end as RawIndex)
+}
+
+pub(crate) fn map_diag<A>(result: anyhow::Result<A>) -> AsmResult<A> {
+    result.map_err(|err| error(loc(0..0), err.to_string()))
+}
+
+// ==========================================================================================
+// Abstract Syntax
+
+/// Represents the AST for an assembly source
+#[derive(Debug)]
+pub struct Unit {
+    /// The name, either script or module.
+    pub name: UnitId,
+    /// A list of address aliases.
+    pub address_aliases: Vec<(Identifier, AccountAddress)>,
+    /// A list of module aliases.
+    pub module_aliases: Vec<(Identifier, ModuleId)>,
+    /// List of functions
+    pub functions: Vec<Fun>,
+}
+
+#[derive(Debug, Clone)]
+pub enum UnitId {
+    Script,
+    Module(ModuleId),
+}
+
+impl UnitId {
+    pub fn module_opt(&self) -> Option<&ModuleId> {
+        match self {
+            UnitId::Script => None,
+            UnitId::Module(id) => Some(id),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Fun {
+    pub loc: Loc,
+    pub name: Identifier,
+    pub visibility: Visibility,
+    pub type_params: Vec<(Identifier, AbilitySet)>,
+    pub params: Vec<Local>,
+    pub locals: Vec<Local>,
+    pub result: Vec<Type>,
+    pub instrs: Vec<Instruction>,
+}
+
+#[derive(Debug)]
+pub enum Type {
+    Named(PartialIdent, Option<Vec<Type>>),
```

Comment:
Would be nice to document these variants, especially since fields are not named. E.g. is bool field in ref mut/immut? etc.